components/director/internal/domain/formation/service.go

```diff
@@ -180,8 +195,189 @@ func (s *service) UnassignFormation(ctx context.Context, tnt, objectID string, o
+	runtimes, err := s.runtimeRepo.ListAll(ctx, in.TargetTenantID, nil)
+
+	if err != nil {
+		return errors.Wrapf(err, "while fetching runtimes in target tenant: %s", in.TargetTenantID)
+	}
+	for _, runtime := range runtimes {
+		if _, err = s.AssignFormation(ctx, in.Tenant, runtime.ID, graphql.FormationObjectTypeRuntime, model.Formation{Name: in.ScenarioName}); err != nil {
+			return errors.Wrapf(err, "while assigning runtime with id %s to formation %s coming from ASA", runtime.ID, in.ScenarioName)
+		}
+	}
+	return nil
+}
+
+// RemoveAssignedScenario removes all the scenarios that are coming from the provided ASA
+func (s *service) RemoveAssignedScenario(ctx context.Context, in model.AutomaticScenarioAssignment) error {
+	runtimes, err := s.runtimeRepo.ListAll(ctx, in.TargetTenantID, nil)
+	if err != nil {
+		return errors.Wrapf(err, "while fetching runtimes in target tenant: %s", in.TargetTenantID)
+	}
+
+	for _, runtime := range runtimes {
+		if _, err = s.UnassignFormation(ctx, in.Tenant, runtime.ID, graphql.FormationObjectTypeRuntime, model.Formation{Name: in.ScenarioName}); err != nil {
+			return errors.Wrapf(err, "while unassigning runtime with id %s from formation %s coming from ASA", runtime.ID, in.ScenarioName)
+		}
+	}
+	return nil
+}
+
+// RemoveAssignedScenarios removes all the scenarios that are coming from any of the provided ASAs
+func (s *service) RemoveAssignedScenarios(ctx context.Context, in []*model.AutomaticScenarioAssignment) error {
+	for _, asa := range in {
+		if err := s.RemoveAssignedScenario(ctx, *asa); err != nil {
+			return errors.Wrapf(err, "while deleting automatic scenario assigment: %s", asa.ScenarioName)
+		}
+	}
+	return nil
+}
+
+// DeleteManyASAForSameTargetTenant deletes a list of ASAs for the same targetTenant
+// It also removes corresponding scenario assignments coming from the ASAs
+func (s *service) DeleteManyASAForSameTargetTenant(ctx context.Context, in []*model.AutomaticScenarioAssignment) error {
+	tenantID, err := tenant.LoadFromContext(ctx)
+	if err != nil {
+		return err
+	}
+
+	targetTenant, err := s.ensureSameTargetTenant(in)
+	if err != nil {
+		return errors.Wrap(err, "while ensuring input is valid")
+	}
+
+	if err = s.repo.DeleteForTargetTenant(ctx, tenantID, targetTenant); err != nil {
+		return errors.Wrap(err, "while deleting the Assignments")
+	}
+
+	if err = s.RemoveAssignedScenarios(ctx, in); err != nil {
+		return errors.Wrap(err, "while unassigning scenario from runtimes")
+	}
+
+	return nil
+}
+
+// MergeScenariosFromInputLabelsAndAssignments merges all the scenarios that are part of the resource labels (already added + to be added with the current operation)
+// with all the scenarios that should be assigned based on ASAs.
+func (s *service) MergeScenariosFromInputLabelsAndAssignments(ctx context.Context, inputLabels map[string]interface{}, runtimeID string) ([]interface{}, error) {
+	scenariosSet := make(map[string]struct{})
```

Comment:
Define `scenariosSet` after `scenariosFromAssignments` and with capacity = len(scenariosFromAssignments).